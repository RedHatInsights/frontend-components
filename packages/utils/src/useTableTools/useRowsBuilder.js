import { emptyRows } from './NoResultsTable';

const columnProp = (column) => {
    return column.key || column.original?.toLowerCase() || column.title?.toLowerCase();
};

const itemRow = (item, columns) => ({
    ...item.rowProps,
    itemId: item.itemId,
    cells: columns.map((column) => ({
        title: column.renderFunc ? column.renderFunc(item) : item[columnProp(column)]
    }))
});

const primeItem = (item, transformers) => {
    let newItem = item;

    transformers.forEach((transformer) => {
        if (transformer) {
            newItem = transformer(newItem);
        }
    });

    return newItem;
};

const applyTransformers = (items, transformers = []) => (
    items.map((item) => (
        primeItem(item, transformers)
    ))
);

const buildRows = (items, columns, options) => {
    const EmptyRowsComponent = options.emptyRows || emptyRows;
    const transformedItems = options?.transformers ?
        applyTransformers(items, options?.transformers) : items;

    const filteredItems = options?.filter ?
        options.filter(transformedItems) : items;

    const sortedItems = options?.sorter ?
        options.sorter(filteredItems) : filteredItems;

    const paginatedItems = options?.paginate?.paginator ?
        options?.paginate?.paginator(sortedItems) : sortedItems;

    const rows = paginatedItems.length > 0 ? paginatedItems.map((item) => (
        itemRow(item, columns)
    )).filter((v) => (!!v)) : EmptyRowsComponent;

    const pagination = options?.paginate?.pagination ? {
        ...options.paginate.pagination,
        itemCount: filteredItems.length
    } : undefined ;

    return {
        rows,
        pagination
    };
};

const useRowsBuilder = (columns, options) => (
    (items) => (
        buildRows(items, columns, options)
    )
);

export default useRowsBuilder;
