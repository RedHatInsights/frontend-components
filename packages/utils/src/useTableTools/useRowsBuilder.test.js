import { renderHook } from '@testing-library/react-hooks';
import useRowsBuilder from './useRowsBuilder';

describe('useRowsBuilder', () => {
    it('returns a rows configuration', () => {
        const { result } = renderHook(() =>
            useRowsBuilder([{ name: 'Item #1' }])
        );
        expect(result).toMatchSnapshot();
    });
});
