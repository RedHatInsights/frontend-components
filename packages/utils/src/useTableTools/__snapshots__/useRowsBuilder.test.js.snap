// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRowsBuilder returns a rows configuration 1`] = `
Object {
  "all": Array [
    [Function],
  ],
  "current": [Function],
  "error": undefined,
}
`;
